### ISSUE TITLE
[Feature Request] Apply `detach` to `y_pred` in metrics

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#221]
Following the discussion in slack:

vfdev : 
In some sense currently we have probably a protection as metrics are computed in 
evaluator that calls with `torch.no_grad()` so there is no graph built 
even after in metric's update methods. However in general case we need 
to reconsider that... 

Jason Kriss :
Ah yea, I guess since we do support accumulating metrics on a trainer (and not just an evaluator) we need to handle this.

**Comment by alykhantejani** [comment]
I must have missed this discussion, can one of you briefly describe the problem +  the proposed solution? 

cc @vfdev-5 @jasonkriss

**Comment by vfdev-5** [comment]
@alykhantejani I was just wondering whether we should not apply `detach` to `y_pred` in metrics before computing some cumulative variables. I was trying to add `EpochMetric` and when `y_pred` are cumulated without detaching we update the graph etc.. And actually this is the case

**Comment by jasonkriss** [comment]
Just to add on to what @vfdev-5 is saying... this really only applies when calculating metrics in a training engine (where there is no call to `torch.no_grad`).

**Comment by anmolsjoshi** [comment]
Is help wanted for this issue? Can I help?

**Comment by vfdev-5** [comment]
> Is help wanted for this issue? Can I help?

@anmolsjoshi Thanks ! Yes, you can go ahead with this issue. The idea is to incorporate `torch.no_grad` somewhere in `Metrics` a "less intrusive" manner.

**Comment by anmolsjoshi** [comment]
@vfdev-5 I wanted to clarify a few things before I got started.

Is this needed for situations where metrics are calculated on each iteration during the training process? And not at the end of an epoch as shown [here](https://github.com/pytorch/ignite/blob/master/examples/mnist/mnist.py#L77-L84)

If so, do you have an example of using Metrics in that way?

**Comment by anmolsjoshi** [comment]
@vfdev-5 I tried a few different methods to incorporate torch.no_grad in Metrics. I thought about creating a new function in Metrics, which detached y_pred, it would work as shown below. 

```python
def iteration_completed(self, engine):
        output = self._detach_output(self._output_transform(engine.state.output))
        self.update(output)

def detach_output(self, output):
        y_pred, y = output
        y_pred = y_pred.detach()
        return y_pred, y
```
Any thoughts?

This method worked well, but I'm still going to look into other options.

**Comment by vfdev-5** [comment]
@anmolsjoshi i think you can try to use decorator `torch.no_grad()` on update metric function...  https://pytorch.org/docs/stable/autograd.html#locally-disable-grad

 And check which metric class really accumulates tensors with grads.

**Comment by anmolsjoshi** [comment]
@vfdev-5 I'm looking into this more, but I did try a few variations of that before the above suggested method. Still looking into it. Here is a simple example showing that it may not work as hoped.

```python 

from abc import ABCMeta, abstractmethod
import torch

class Metric(object):
    __metaclass__ = ABCMeta

    def __init__(self):
        self.num = 0

    @abstractmethod
    @torch.no_grad()
    def update(self, output):
        pass

class ExampleMetric(Metric):
    def update(self, output):
        y_pred, y = output
        mul = 2*y_pred
        print (mul)

y = torch.zeros(4, requires_grad=False)
y_pred = torch.zeros(4, requires_grad=True)
outputs = [y_pred, y]

metric = ExampleMetric()
metric.update(outputs)
```

Here is the output - `tensor([0., 0., 0., 0.], grad_fn=<MulBackward>)`

So despite torch.no_grad() being called on update in Metric, it still results in an output that has a grad_fn. I think this is because @torch.no_grad() is overridden when update is defined again in ExampleMetric. 

I'm still looking into this.

**Comment by vfdev-5** [comment]
@anmolsjoshi nice example! Actually, I'm not against applying `@torch.no_grad` at every derived class:
```python
from abc import ABCMeta, abstractmethod
import torch

class Metric(object):
    __metaclass__ = ABCMeta

    def __init__(self):
        self.num = 0

    @abstractmethod
    def update(self, output):
        pass

class ExampleMetric(Metric):
    
    @torch.no_grad()    
    def update(self, output):
        y_pred, y = output
        mul = 2*y_pred
        print (mul)

y = torch.zeros(4, requires_grad=False)
y_pred = torch.zeros(4, requires_grad=True)
outputs = [y_pred, y]

metric = ExampleMetric()
metric.update(outputs)
```
you're right `update` is overridden and base method is not called and decorator is not applied.

More proper design would require to introduce `_update` : 
```
from abc import ABCMeta, abstractmethod
import torch

class Metric(object):
    __metaclass__ = ABCMeta

    def __init__(self):
        self.num = 0

    @torch.no_grad()
    def update(self, output):
        return self._update(output)

    @abstractmethod
    def _update(self, output):
        pass

class ExampleMetric(Metric):
    
    def _update(self, output):
        y_pred, y = output
        mul = 2*y_pred
        print (mul)

y = torch.zeros(4, requires_grad=False)
y_pred = torch.zeros(4, requires_grad=True)
outputs = [y_pred, y]

metric = ExampleMetric()
metric.update(outputs)
```

**Comment by vfdev-5** [comment]
@anmolsjoshi no, I let you submit a PR with the code and with tests, please, if you do not mind :)

**Comment by anmolsjoshi** [comment]
@vfdev-5 thanks for the opportunity!

**Comment by vfdev-5** [comment]
Closed by #236
